include/zephyr/senss/senss.h

```diff
@@ -0,0 +1,384 @@
+ *
+ * Reporting sensor data when the sensor data changes.
+ *
+ * Exclusive with  \ref SENSS_SENSOR_FLAG_REPORT_ON_EVENT
+ */
+#define SENSS_SENSOR_FLAG_REPORT_ON_CHANGE			BIT(1)
+
+
+/**
+ * @brief Sensor state.
+ *
+ */
+enum senss_sensor_state {
+	SENSS_SENOSR_STATE_NOT_READY = 1,
+	SENSS_SENOSR_STATE_READY = 2,
+};
+
+
+/**
+ * @struct senss_sensor_info
+ * @brief Sensor basic constant information
+ *
+ */
+struct senss_sensor_info {
+	/** Name of the sensor instance */
+	const char *name;
+
+	/** Friendly name of the sensor instance */
+	const char *friendly_name;
+
+	/** Vendor name of the sensor instance */
+	const char *vendor;
+
+	/** Model name of the sensor instance */
+	const char *model;
+
+	/** Sensor type */
+	int32_t type;
+
+	/** Sensor index for multiple sensors under same sensor type */
+	int32_t sensor_index;
+
+	/** Sensor flags */
+	uint32_t flags;
+
+	/** Minimal report interval in micro seconds  */
+	uint32_t minimal_interval;
+
+	/** Sensor version */
+	struct senss_sensor_version version;
+};
+
+
+
+/**
+ * @brief Initialize sensor subsystem runtime.
+ *
+ * This will allocate internal resources, such as create internal schedule
+ * task, create internal management data structures.
+ *
+ * It will also enumerate all registered sensor instances in sensor subsystem,
+ * build sensors report relational data model, if a virtual sensor's required dependency
+ * sensor type not found, will return error.
+ * Then call each sensor instances' init callback API.
+ *
+ * @return 0 on success or negative error value on failure.
+ *
+ */
+int senss_init(void);
+
+ /**
+  * @brief De-initialize sensor subsystem runtime.
+  *
+  * This will release all dynamically allocated internal resources, and call
+  * all sensor instances' deinit callback API.
+  *
+  * @return 0 on success or negative error value on failure.
+  */
+int senss_deinit(void);
+
+ /**
+  * @brief Get all supported sensor instances' information.
+  *
+  * This API just returned read only information of sensor instances,
+  * no side effect to sensor instances.
+  *
+  * @param **info For receiving sensor instances' information array pointer.
+  *
+  * @return Number of sensor instances, 0 returned if no sensors.
+  */
+int senss_get_sensors(const struct senss_sensor_info **info);
+
+
+/**
+ * @brief Open sensor instance by sensor type and sensor index.
+ *
+ * Application clients use it to open a sensor instance and get its handle.
+ * Support multiple Application clients for open same sensor instance,
+ * in this case, the returned handle will different for different clients.
+ *
+ * @param type The sensor type which need to open.
+ *
+ * @param sensor_index The index which need to open.
+ *
+ * @param *handle The opened instance handle, if failed will be set to -1.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_open_sensor(
+		int type, int sensor_index,
+		int *handle);
```

Comment:
Can you explain why we need to "open/close" each sensor? Can we not just start/stop a client which gets us a client ID then use the client ID for transactions with sensors?

For example:

```
int client_id;
int rc = ssense_open(&client_id);

__ASSERT(rc == 0);

size_t num_sensors;
const struct ssense_sensor_info *sensors;

rc = ssense_get_sensors(&sensors, &num_sensors);
__ASSERT(rc == 0);

size_t selected_sensor;
for (size_t i = 0; i < num_sensors; ++i) {
  if (sensors[i]->type == SENSOR_TYPE_ACCEL) {
    selectd_sensor = i;
    break;
  }
}

rc = ssense_set_interval(client_id, sensor_index, 1000);
```

The combination of client_id and sensor_index should be unique enough to do arbitration

Reply:
I would also recommend using a typedef for the client ID in case we ever need to change it.

Reply:
hi, yperess, thanks for your suggestion. let me explain more why we use open/close method for client/report in senss: 
1) Our senss design model is: take sensor as a point, handle is taken as a connection between reporter sensor and client sensor. In senss.c senss_close_sensor() function, **struct senss_connection *conn = get_connection_by_handle(ctx, handle);**, we will transfer handle to connection, and pass it to sensor_mgmt.c, all sensor relationship processing is based on this connection.
2)  This connection is not only apply to application client(created by open_sensor()), but also apply to client and reporter defined by device tree. So senss_set_interval() could be called by any connection.
If using client_id and sensor_index as you referred, we need to modify sensor_mgmt implementation, it will introduce huge of work, and also this model maybe called another sensor model, but not sensor subsystem model.

Reply:
@ghu0510 it seems that we touch on the same point in a lot of places during this review so I'll focus the conversation here for now to keep things simpler.

I understand what you mean by _connection_ here but I think it's not needed. The pair of `client_id` and `sensor_index` is enough for any unique pairing. It also prevents you from accidentally "opening" too many sensor connections. Effectively, every 1 client is allowed to interact with every sensor once. So as an example,

**We have 3 clients on chromebooks**
1. The AP which can send host commands to the EC (or ISH as we're working on enabling it) and request sensor data with varying attributes (rate etc)
2. The EC itself to do things like calculate lid angle
3. Eventually we'll also have CHRE as a client

These clients can and should only "open" a single connection to the sensor subsystem.

**What is a connection/client**
A connection has a 0 or 1 callbacks for the following:
* Data events
* Sampling status changes (interval, latency, enabled)
* Bias events
* Flush completion (clearing out the batch)

![client](https://user-images.githubusercontent.com/9535482/225514531-c5244b73-1e5c-4e2a-a408-0c732f8fc2a9.jpeg)

Generally I would prefer for this to be a part of the `open` call, but I'm OK with having separate functions like you have in the PR currently. So I would prefer something like:

```
/**
 * @brief Iterates through 'sensors' until it finds one with type SENSOR_TYPE_ACCEL
 * @param sensors List of sensors from sensor_subsys_get_sensors()
 * @param num_sensors The number of sensors in the 'sensors' list
 * @param idx Output variable for the index of the first accelerometer in the list
 * @return 0 on success, < 0 on error
 */
int find_accelerometer(
    const struct sensor_subsys_sensor_info *sensors,
    size_t num_sensors,
    size_t *idx
);

/** Same as find_accelerometer but for SENSOR_TYPE_MAG */
int find_magnetometer(
    const struct sensor_subsys_sensor_info *sensors,
    size_t num_sensors,
    size_t *idx
);

static struct sensor_subsys_client_api my_client = {
  .on_sample_status_change = my_on_sample_status_change,
  .on_data_event = my_on_data_event,
  .on_bias_event = NULL,
  .on_flush_complete = NULL,
};

int main(void) {
  const struct sensor_subsys_sensor_info *sensors;
  size_t num_sensors;
  size_t accel_idx;
  size_t mag_idx;

  sensor_subsys_client_id_t client_id = sensor_subsys_open(&my_client);

  sensor_subsys_get_sensors(&sensors, &num_sensors);
  find_accelerometer(sensors, num_sensors, &accel_idx);
  find_magnetometer(sensors, num_sensors, &mag_idx);

  struct sensor_subsys_sensor_config config = {
    .mode = SENSOR_SUBSYS_CONFIG_MODE_CONTINUOUS,
    .interval_ns = UINT64_C(31250), // 32kHz
    .latency_ns = UINT64_C(1000000), // 1ms batching
  };

  // Once we configure the sensor for this client we expect 2 things to start happening
  // 1. my_on_sample_status_change() should be called
  // 2. my_on_data_event() will be called every 1ms
  sensor_subsys_configure(client_id, accel_idx, &config);

  config.interval_ns = UINT64_C(20000000); // 50Hz
  sensor_subsys_configure(client_id, mag_idx, &config);
}
```

This approach does a few things:
1. It keeps the clients 1:1 with their callbacks opening a new client is tied to the client's callbacks.
2. Setting the various attributes separately is difficult. The existing approach on this PR has us setting interval and latency separately. What happens if we set the interval first, then context switch, get some samples, then come back to our thread, then set the latency. While it probably doesn't break anything, it might as we get into more detailed attributes like scale
3. The application now only manages 1 connection. Note that all the data will be sent to the same handler `my_on_data_event()` instead of a separate callback for every sensor. The idea is to separate the client from the sensor configuration. Lets assume an application that allows 4 external clients (I use the word external since we can calculate at build time how many virtual sensors we have from devicetree so we can allocate the client buffer to that + some extras for runtime) and has 7 sensors (lets say 5 physical and 2 virtual). In this example we would need to allocate room for 6 clients (4 external and 2 virtual sensors) and each client can request different configurations for each sensor so that gives us `4 * 7` for the external connections and `2 * 6` for the virtual sensors (since virtual sensors can't be their own clients (40 possible configurations). If each client + sensor is considered a new connection (like in this PR) that means we need to allocate client metadata + configuration metadata for each sensor. Alternatively, if we separate the 2 concepts, we only need client metadata for 6 clients and 40 configurations (saving us 34 entries for client metadata). Now this might not seem like a big deal but there are 2 considerations that need to be made
    1. most boards don't have the large capacity of the ISH
    2. because this is going in Zephyr we need to consider much more than just laptops. If this subsystem is used in a robotics application you might get many many more sensors, now having a connection for every single client/sensor pair grows quadratically instead of linearly.

Reply:
@yperess, appreciate for your comments, many ideas that valuable for discussion, let's discuss one by one. 😄   

> @ghu0510 it seems that we touch on the same point in a lot of places during this review so I'll focus the conversation here for now to keep things simpler.
> 
> I understand what you mean by _connection_ here but I think it's not needed. The pair of `client_id` and `sensor_index` is enough for any unique pairing. It also prevents you from accidentally "opening" too many sensor connections. Effectively, every 1 client is allowed to interact with every sensor once. So as an example,
> 
> **We have 3 clients on chromebooks**
> 
> 1. The AP which can send host commands to the EC (or ISH as we're working on enabling it) and request sensor data with varying attributes (rate etc)
> 2. The EC itself to do things like calculate lid angle
> 3. Eventually we'll also have CHRE as a client
> 
> These clients can and should only "open" a single connection to the sensor subsystem.
> 
> **What is a connection/client** A connection has a 0 or 1 callbacks for the following:
> 
> * Data events
> * Sampling status changes (interval, latency, enabled)
> * Bias events
> * Flush completion (clearing out the batch)
> 
> ![client](https://user-images.githubusercontent.com/9535482/225514531-c5244b73-1e5c-4e2a-a408-0c732f8fc2a9.jpeg)
> 
> Generally I would prefer for this to be a part of the `open` call, but I'm OK with having separate functions like you have in the PR currently. So I would prefer something like:
> 
> ```
> /**
>  * @brief Iterates through 'sensors' until it finds one with type SENSOR_TYPE_ACCEL
>  * @param sensors List of sensors from sensor_subsys_get_sensors()
>  * @param num_sensors The number of sensors in the 'sensors' list
>  * @param idx Output variable for the index of the first accelerometer in the list
>  * @return 0 on success, < 0 on error
>  */
> int find_accelerometer(
>     const struct sensor_subsys_sensor_info *sensors,
>     size_t num_sensors,
>     size_t *idx
> );
> 
> /** Same as find_accelerometer but for SENSOR_TYPE_MAG */
> int find_magnetometer(
>     const struct sensor_subsys_sensor_info *sensors,
>     size_t num_sensors,
>     size_t *idx
> );
> ```

Since there would be tens of sensor types, so, I would like to define a single API instead of each type has a `find_xxx` API, for example: 

```c
/**
 * @brief Find sensor instances by sensor type and sensor index.
 *
 * Application can use these API to find sensor instances' information by
 * sensor type and sensor index, or enumerate all sensor instances'
 * information at one call.
 *
 * This API just returned read only information of sensor instances,
 * no side effect to sensor instances.
 *
 * @param type The sensor type which need to find,
 * \ref SENSS_SENSOR_TYPE_ALL for all types.
 *
 * @param sensor_index The index which need to find, -1 for find all
 * instances with type \p type.
 *
 * @param *info Input info array for receiving found sensor instances
 *
 * @param max The max count of the \p infos array input. Can get real count
 * number via \ref senss_get_sensor_count.
 *
 * @return Number of sensor instances foun
 */
int senss_find_sensor(
  	int type, int sensor_index,
  	struct senss_sensor_info *info, int max);
```

And if we let user to allocate buffer to receive find results according to real sensor instance number to avoid waste memory, we can define an API to get the sensor instance number: 
                                       
```c
int senss_get_sensor_count(
  	int type, int sensor_index);
```

Or if let the find API to allocate the results buffer, can just have one API that simpler: 

```c
int senss_find_sensor(
  	int type, int sensor_index,
  	struct senss_sensor_info *info, int* sensor_number);
```

I think we already have many discussions about this in https://github.com/zephyrproject-rtos/zephyr/issues/48149 
> ```
> static struct sensor_subsys_client_api my_client = {
>   .on_sample_status_change = my_on_sample_status_change,
>   .on_data_event = my_on_data_event,
>   .on_bias_event = NULL,
>   .on_flush_complete = NULL,
> };
> 
> int main(void) {
>   const struct sensor_subsys_sensor_info *sensors;
>   size_t num_sensors;
>   size_t accel_idx;
>   size_t mag_idx;
> 
>   sensor_subsys_client_id_t client_id = sensor_subsys_open(&my_client);
> 
>   sensor_subsys_get_sensors(&sensors, &num_sensors);
>   find_accelerometer(sensors, num_sensors, &accel_idx);
>   find_magnetometer(sensors, num_sensors, &mag_idx);
> 
>   struct sensor_subsys_sensor_config config = {
>     .mode = SENSOR_SUBSYS_CONFIG_MODE_CONTINUOUS,
>     .interval_ns = UINT64_C(31250), // 32kHz
>     .latency_ns = UINT64_C(1000000), // 1ms batching
>   };
> 
>   // Once we configure the sensor for this client we expect 2 things to start happening
>   // 1. my_on_sample_status_change() should be called
>   // 2. my_on_data_event() will be called every 1ms
>   sensor_subsys_configure(client_id, accel_idx, &config);
> 
>   config.interval_ns = UINT64_C(20000000); // 50Hz
>   sensor_subsys_configure(client_id, mag_idx, &config);
> }
> ```

    

> 
> This approach does a few things:
> 
> 1. It keeps the clients 1:1 with their callbacks opening a new client is tied to the client's callbacks.

In sometimes, user may need separate callbacks per sensor, this PR provide the most flexibly for user to use single callbacks for all sensor or separate callbacks for each sensor. 

> 2. Setting the various attributes separately is difficult. The existing approach on this PR has us setting interval and latency separately. What happens if we set the interval first, then context switch, get some samples, then come back to our thread, then set the latency. While it probably doesn't break anything, it might as we get into more detailed attributes like scale

Use a common `sensor_subsys_sensor_config` with combined config parameters, can sue single API to configure All parameters, and may avoid the context switch situation you mentioned, but it has other problems:
- All parameters in single config data structure or partially with others still go separate config API?
  - If all, will have forward compatible issue, if add new config parameters in future, need add to the single common config data structure, this will break existing code which no need these new parameters.
  - If partially, it's confused for different API style, not good.     
- Not easy use for user especially if need frequently change some parameters but keep others.
  - User need maintain the single config data structure, change parameter(s) they need, and call set config API
  - Set config API internally need monitor which parameter(s) changes and do related side effect.
  - User need fill the default value of no interesting parameter data fields, if with incorrect value will cause un-expected behavior, if use separate config APIs, will have no such issue since for no interesting parameters just not call their config APIs.      
- If reuse the config data structure that user allocated
  - **Good**
    - saving memory
  - **Bad**
    - **dangerous** if user allocates the config data structure in stack
    - hard to track the parameters changes in the config data structure, for example if user changed the value of `interval_ns` intently or unwittingly, sensor subsystem doesn't know this, this will break the logic in sensor subsystem. 
- If copy the config data structure that user allocated, that do not save memory.  

For the context switch situation you mentioned, I don't think it's a problem, in most time, it doesn't matter, if user really care about the changing stage, user should stop the streaming at first, then config the new parameters, at last re-start the sensor.      
 
> 3. The application now only manages 1 connection. Note that all the data will be sent to the same handler `my_on_data_event()` instead of a separate callback for every sensor. The idea is to separate the client from the sensor configuration. Lets assume an application that allows 4 external clients (I use the word external since we can calculate at build time how many virtual sensors we have from devicetree so we can allocate the client buffer to that + some extras for runtime) and has 7 sensors (lets say 5 physical and 2 virtual). In this example we would need to allocate room for 6 clients (4 external and 2 virtual sensors) and each client can request different configurations for each sensor so that gives us `4 * 7` for the external connections and `2 * 6` for the virtual sensors (since virtual sensors can't be their own clients (40 possible configurations). If each client + sensor is considered a new connection (like in this PR) that means we need to allocate client metadata + configuration metadata for each sensor. Alternatively, if we separate the 2 concepts, we only need client metadata for 6 clients and 40 configurations (saving us 34 entries for client metadata). Now this might not seem like a big deal but there are 2 considerations that need to be made
>    
>    1. most boards don't have the large capacity of the ISH
>    2. because this is going in Zephyr we need to consider much more than just laptops. If this subsystem is used in a robotics application you might get many many more sensors, now having a connection for every single client/sensor pair grows quadratically instead of linearly.

The two options of client connection: 
- **A: subsystem level connection (your suggestion)**
- **B: per sensor connection (this PR)**
 
**option A** is still need handle per sensor configuration and arbitration, compared with **option B**, it can only save external client/sensor map data per sensor, which is two pointers with total 8 Bytes in a 32bit system, with your example it saves total 4 * 7 * 8 = 224 Bytes, other per sensor metadata including config data and runtime state data also needed and can't be reduced, these parts are the big part. 

To me, **option A** can save some memory but limited. Even for an example robotics application with many sensors, such as 100 sensors, with 4 application clients and each client will access all the sensors, **option A** can save 4 * 100 * 8 = 3200 Bytes. But if a system can have 100 sensors and with 4 applications use all of these 100 sensors, I don't think a low-end MCU can do this, so in this case, saving 3200 Bytes for such system is not so valuable. 

My understand of **option A** can save memory because it assumes the application client(s) will access all the sensors from the subsystem and it saves the per sensor client map data. But if an application just need access few sensors of the subsystem, how does the sensor subsystem to allocate per sensor metadata (config data may allocated by user according needs, how about the runtime state data), pre allocate for all sensors will waste of memory. 

The `senss_open/senss_close` in **option B** provide the explicit APIs for application client(s) to open/close (allocate/release resource) only their interesting sensors, avoid allocate no needed metadata for no interesting sensors.         
               
Sensor subsystem can have a super set of sensors, different applications can just open their interesting sensors, which maybe a subset of the sensors, for example the CHRE client may not access all the sensors, not all sensors need expose to the AP for host link client, and a native application client may only access a Acc sensor. I think this is a very common usage. 

So, I still prefer **option B** 😃

Reply:
@yperess, it's more clear now, thanks for your clarifications!  
> 
> > > ```
> > > int find_accelerometer(
> > >     const struct sensor_subsys_sensor_info *sensors,
> > >     size_t num_sensors,
> > >     size_t *idx
> > > );
> > > 
> > > /** Same as find_accelerometer but for SENSOR_TYPE_MAG */
> > > int find_magnetometer(
> > >     const struct sensor_subsys_sensor_info *sensors,
> > >     size_t num_sensors,
> > >     size_t *idx
> > > );
> > > ```
> 
> This API was meant to be a sample of what the application might write to iterate through the `senss_sensor_info` structs and find the accel/mag they wanted. They might have different parameters for which accel they want (such as resolution etc). I wasn't suggesting we should implement them. The thinking is that the application calls `sensss_get_sensors` then does whatever they want to find which sensor they want.
> 

Thanks!  I see.

> > In sometimes, user may need separate callbacks per sensor, this PR provide the most flexibly for user to use single callbacks for all sensor or separate callbacks for each sensor.
> 
> I don't think they ever really _need_ a separate callback. If we give 1 callback for each connection they can always switch off the type or the sensor ID for their custom handling logic.
> 

Yes, I agree, and I don't think single callback or separate callbacks are critical, just different styles. For single callback, user no need multipe registration of callbacks, but need write many switch of cases with type or id, while with separate callbacks, user need multipe registration, but no need writes many switch of cases. It's hard to see which style is better. **option B** just provides the flexibility for these two different styles. :）  

> > > 2. Setting the various attributes separately is difficult. The existing approach on this PR has us setting interval and latency separately. What happens if we set the interval first, then context switch, get some samples, then come back to our thread, then set the latency. While it probably doesn't break anything, it might as we get into more detailed attributes like scale
> > 
> > 
> > Use a common `sensor_subsys_sensor_config` with combined config parameters, can sue single API to configure All parameters, and may avoid the context switch situation you mentioned, but it has other problems:
> 
> Another approach to consider is what we'll be proposing for the kernel drivers API where the user can give a list of attributes to update. It looks like:
> 
> ```c
> struct senss_config {
>   enum senss_attribute attribute;
>   union {
>     int32_t int_val;
>     bool bool_val;
>     ...
>   };
> }
> 
> int senss_configure_sensor(client_id, sensor_idx, struct senss_config *configs, size_t count);
> ```
> 

Yeah, this approach is good for large amount of attributes cases and friendly for implementation, but not friendly for users, and I don't think our sensor applications will have so many attributes, so, I still prefer the simple and easy used separate config APIs.   

> > To me, **option A** can save some memory but limited.
> 
> Sure, the memory savings are limited and you're 100% right. But I'll argue that it simplifies the interaction with the subsystem. Having 1 connection makes the lookup easier in 99% of the use cases. Under **option B** when the AP requests an update to sensor '2'. We currently have a mapping function on the chromium EC [motion_sense.c;l=939](https://source.chromium.org/chromium/chromiumos/platform/ec/+/main:common/motion_sense.c;l=939;drc=22fd582040864730ca19d4c72a404c38f6e02267) which is possible because the AP gets the list from the EC. If we move to a per connection setup, I believe we would need to introduce a lookup table of the host sensor index to a connection handle. This will again be true for our PAL implementation for CHRE and any other client of the subsystem. So my concern is that we're making every client do the same thing. Generally when that's the case, I find that we can probably come up with a simpler solution and handle it for the clients instead.
> 

If `senss` is only for chre, I totally agree with you that **option A** is better. But we can't assume all application clients all use host sensor IDs like Chrome OS, for example the `HID` on Windows used `report ID` to identify different sensor reports, `report ID` counts from `1` and under a `TLC` (Top Level Collection), and a sensor hub can expose multiple `TLCs`, in this case, `report ID` can't directly use as sensor index to `senss`, we still need a map.   

And for the map you mentioned in [motion_sense.c;l=939](https://source.chromium.org/chromium/chromiumos/platform/ec/+/main:common/motion_sense.c;l=939;drc=22fd582040864730ca19d4c72a404c38f6e02267), is just a simple data object convert like: `sensor = &motion_sensors[host_id]; `, should have no performance impacts, and should not a problem. 

> **option B** is more explicit and thus has more room for error. You've mentioned (rightfully) that having a single config struct can lead to errors made by careless developers. Same can be said for letting them manage each individual connection. I don't think it'll be hard for us to manage these connections inside the subsystem. We would have an array of pre-allocated connection structs (hidden from the client). When a request comes in to configure a `client/sensor` pair we would basically see if one was already allocated. So internally we can still keep things per connection like you have, I just don't see the need to expose that to the client.

**option B** is more explicit and easier for user to just access some few sensors, for example, some application may just want access an Accelerometer sensor, can just use `senss_open_sensor(SENSS_SENSOR_TYPE_MOTION_ACCELEROMETER_3D, 0)` which is easier and nature for user. `senss_open_sensor()` also provides a right place to allocate internal resource (per sensor internal runtime metadata for the client) for just the accessed sensor. **option A** can't know which sensors will be accessed by user, without APIs like `senss_open_sensor()`, internally sensor subsystem don't know how many sensor's per client runtime metadata need allocate, it's good for applications need access all the sensors in sensor subsystem.

Reply:
> Yes, I agree, and I don't think single callback or separate callbacks are critical, just different styles. For single callback, user no need multipe registration of callbacks, but need write many switch of cases with type or id, while with separate callbacks, user need multipe registration, but no need writes many switch of cases. It's hard to see which style is better. **option B** just provides the flexibility for these two different styles. :）

I this this topic converges with the discussion below for 1 connection per client vs 1 connection per sensor. So I'm going to comment below and leave this thread in order to keep the review simpler.
  
> Yeah, this approach is good for large amount of attributes cases and friendly for implementation, but not friendly for users, and I don't think our sensor applications will have so many attributes, so, I still prefer the simple and easy used separate config APIs.

I see your point, but I think that having separate APIs for every configurable attribute just doesn't scale. I think your point is that it's easier not to set up all the configs in an array of structs and it feels much more native to have a separate function call per. Would the following work for what you're looking for?

```
int senss_sensor_configure(
    senss_client_id_t client_id,
    size_t sensor_index,
    const struct senss_sensor_config *configs,
    size_t count
);

/**
 * Example use:
 *   senss_sensor_set_config(client_id, idx, (struct senss_sensor_config) {
 *     .attribute = SENSS_ATTR_SAMPLE_RATE,
 *     .int_val = 500,
 *   });
 */
static int senss_sensor_set_config(
    senss_client_id_t client_id,
    size_t sensor_index,
    struct senss_sensor_config config
) {
  return senss_sensor_configure(client_id, sensor_index, &config, 1);
}
```

> If `senss` is only for chre, I totally agree with you that **option A** is better. But we can't assume all application clients all use host sensor IDs like Chrome OS, for example the `HID` on Windows used `report ID` to identify different sensor reports, `report ID` counts from `1` and under a `TLC` (Top Level Collection), and a sensor hub can expose multiple `TLCs`, in this case, `report ID` can't directly use as sensor index to `senss`, we still need a map.
> 
> And for the map you mentioned in [motion_sense.c;l=939](https://source.chromium.org/chromium/chromiumos/platform/ec/+/main:common/motion_sense.c;l=939;drc=22fd582040864730ca19d4c72a404c38f6e02267), is just a simple data object convert like: `sensor = &motion_sensors[host_id]; `, should have no performance impacts, and should not a problem.
> 
> > ...
> 
> **option B** is more explicit and easier for user to just access some few sensors, for example, some application may just want access an Accelerometer sensor, can just use `senss_open_sensor(SENSS_SENSOR_TYPE_MOTION_ACCELEROMETER_3D, 0)` which is easier and nature for user. `senss_open_sensor()` also provides a right place to allocate internal resource (per sensor internal runtime metadata for the client) for just the accessed sensor. **option A** can't know which sensors will be accessed by user, without APIs like `senss_open_sensor()`, internally sensor subsystem don't know how many sensor's per client runtime metadata need allocate, it's good for applications need access all the sensors in sensor subsystem.

I think I'm seeing what the misunderstanding is, the vision is to have a client pool and a sensor pool as 2 separate pools. Generally speaking the application should know statically how many clients it'll have (AP, CHRE, etc). So this can be handled with a simple `CONFIG_SENSS_MAX_NUM_CLIENTS` (internally we'll actually allocate `CONFIG_SENSS_MAX_NUM_CLIENTS + DT_NUM_INST_STATUS_OKAY(zephyr_senss_virtual_sensor)`.  These internal structures simply map a `client_id` to its callbacks, the second sensor configuration pool maintains the same information as you do here in the `open` API. So really the only difference is where the callbacks live.

For the sake of moving forward can we compromise on something like:
```
senss_handle_t senss_sensor_open(
    size_t sensor_idx,
    const struct senss_callbacks *callbacks
);

int senss_config_sensor_configure(
    senss_handle_t handle, 
    const struct senss_sensor_config *configs,
    size_t count
);
```

This would keep the semantics of 1 connection per sensor/client pairing that's in the current proposal but also keep the API space smaller by having a single call that opens the connection and sets up the callbacks (we'll need 3 that I can think of right now):
1. Data event
2. Bias event
3. Config changed